src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +121,136 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
             return FromState(newState);
         }
 
+        public GeneratorDriverRunResult GetRunResult()
+        {
+            var results = _state.Generators.ZipAsArray(
+                            _state.GeneratorStates,
+                            (generator, generatorState)
+                                => new GeneratorRunResult(generator,
+                                                          diagnostics: generatorState.Diagnostics,
+                                                          exception: generatorState.Exception,
+                                                          generatedSources: generatorState.SourceTexts.ZipAsArray(
+                                                                                generatorState.Trees,
+                                                                                (sourceText, tree) => new GeneratedSourceResult(tree, sourceText.Text, sourceText.HintName))));
+            return new GeneratorDriverRunResult(results);
+        }
+
+        internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, DiagnosticBag? diagnosticsBag, CancellationToken cancellationToken = default)
+        {
+            // with no generators, there is no work to do
+            if (_state.Generators.IsEmpty)
+            {
+                return _state;
+            }
+
+            // run the actual generation
+            var state = StateWithPendingEditsApplied(_state);
+            var stateBuilder = ArrayBuilder<GeneratorState>.GetInstance(state.Generators.Length);
+            var walkerBuilder = ArrayBuilder<GeneratorSyntaxWalker?>.GetInstance(state.Generators.Length, fillWithValue: null); // we know there is at max 1 per generator
+            int receiverCount = 0;
+
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = state.GeneratorStates[i];
+
+                // initialize the generator if needed
+                if (!generatorState.Info.Initialized)
```

Comment:
>generatorState.Info.Initialized [](start = 21, length = 31)

When would a generator have been initialized?

Reply:
On the next run through. We'll return a drive which contains a set of initialized generators. The IDE will use that to do another generation on a different compilation, and we won't re-init (by design)

---
In reply to: [467288683](https://github.com/dotnet/roslyn/pull/46296#discussion_r467288683) [](ancestors = 467288683)